src/frontend/src/expr/utils.rs

```diff
@@ -587,12 +583,174 @@ impl WatermarkAnalyzer {
                 // TODO: do we need derive watermark when every case can derive a common watermark?
                 WatermarkDerivation::None
             }
-            ExprType::Proctime => Nondecreasing,
+            ExprType::Proctime => WatermarkDerivation::None,
             _ => WatermarkDerivation::None,
         }
     }
 }
 
+#[derive(Clone, Default)]
+pub struct MonotonicityAnalyzer {
+    pub input_mono_cols: Vec<(usize, Monotonicity)>,
+}
+
+impl MonotonicityAnalyzer {
+    pub fn try_derive_monotonicity(&self, expr: &ExprImpl) -> Option<Monotonicity> {
+        self.visit_expr(expr)
+    }
+
+    fn visit_expr(&self, expr: &ExprImpl) -> Option<Monotonicity> {
+        match expr {
+            ExprImpl::InputRef(inner) => self
+                .input_mono_cols
+                .iter()
+                .find_map(|(i, m)| (inner.index == *i).then(|| m.clone())),
+            ExprImpl::Literal(_) => Some(Monotonicity::Constant),
+            ExprImpl::FunctionCall(inner) => self.visit_function_call(inner),
+            ExprImpl::FunctionCallWithLambda(_inner) => None,
+            ExprImpl::TableFunction(_) => None,
+            ExprImpl::Subquery(_)
+            | ExprImpl::AggCall(_)
+            | ExprImpl::CorrelatedInputRef(_)
+            | ExprImpl::WindowFunction(_)
+            | ExprImpl::Parameter(_)
+            | ExprImpl::Now(_) => unreachable!(),
```

Comment:
How do we know these are unreachable? Maybe better to explain a little bit in comment